packages/minimongo/upsert_document.js

```diff
@@ -0,0 +1,123 @@
+        // handle implicit $and
+        for (let [key, value] of Object.entries(query)) {
+            // handle explicit $and
+            if (key === "$and") {
+                value.forEach(sq => populateDocumentWithQueryFields(sq, document))
+            }
+            // handle $or nodes with exactly 1 child
+            else if (key === "$or") {
+                if (value.length === 1) {
+                    populateDocumentWithQueryFields(value[0], document)
+                }
+            }
+            //Ignore other '$'-prefixed logical selectors
+            else if (key[0] !== "$") {
+                populateDocumentWithKeyValue(document, key, value);
+            }
+        }
+    } else {
+        // Handle meteor-specific shortcut for selecting _id
+        if (LocalCollection._selectorIsId(query)) {
+            insertIntoDocument(document, "_id", query);
+        }
+    }
+
+    return document;
+
+}
+
+// Handles one key/value pair to put in the selector document
+function populateDocumentWithKeyValue(document, key, value) {
+    if (value && Object.getPrototypeOf(value) === Object.prototype) {
+        populateDocumentWithObject(document, key, value);
+    } else if (!(value instanceof RegExp)) {
+        insertIntoDocument(document, key, value);
+    }
+}
+
+
+// Handles a key, value to put in the selector document if the value is an object
+function populateDocumentWithObject(document, key, value) {
+    const keys = Object.keys(value);
+    const unprefixedKeys = keys.filter(k => k[0] !== "$");
+
+    if (unprefixedKeys.length > 0 || !keys.length) {
+        //Literal (possibly empty) object ( or empty object ) 
+        //Don't allow mixing '$'-prefixed with non-'$'-prefixed fields
+        if (keys.length !== unprefixedKeys.length) {
+            throw new Error(`unknown operator: ${unprefixedKeys[0]}`)
+        } else {
+            validateObject(value, key);
+            insertIntoDocument(document, key, value);
+        }
+    } else {
+        for (let [k, v] of Object.entries(value)) {
+            if (k === "$eq") {
+                populateDocumentWithKeyValue(document, key, v);
+            }
+            //every value for $all should be dealt with as separate $eq-s
+            else if (k === "$all") {
+                v.forEach(vx => populateDocumentWithKeyValue(document, key, vx))
+            }
+        }
+    }
+}
+
+// Actually inserts a key value into the selector document
+// However, this checks there is no ambiguity in setting the value for the given key, throws otherwise
+function insertIntoDocument(document, key, value) {
+
+    Object.keys(document).forEach(existingKey => {
+        if (
+            (existingKey.length > key.length && existingKey.indexOf(key) === 0)
+            || (key.length > existingKey.length && key.indexOf(existingKey) === 0)
+        ) {
+            throw new Error(`cannot infer query fields to set, both paths '${existingKey}' and '${key}' are matched`)
```

Comment:
Can you make sure longer lines are wrapped at 80 chars? Switching the indentation to 2 chars will help, but some lines (like this one) are still over.